src/main/java/com/redhat/pantheon/model/api/SlingResource.java

```diff
@@ -0,0 +1,274 @@
+package com.redhat.pantheon.model.api;
+
+import org.apache.sling.api.adapter.Adaptable;
+import org.apache.sling.api.resource.*;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.annotation.Nonnull;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.Arrays.stream;
+
+/**
+ * An Editable Sling resource model object. These are simple objects made up of a single constructor based on an existing
+ * sling resource, and a set of fields or child resources mapped to the resource. Fields and child resources are
+ * strongly typed and they can be modified. SlingResources are wrappers around existing sling resources and as such the
+ * resource needs to exist before it can be wrapped by one of these.
+ */
+public class SlingResource implements Adaptable {
+
+    private final Resource resource;
+
+    public SlingResource(@Nonnull Resource resource) {
+        this.resource = resource;
+    }
+
+    public Resource getResource() {
+        return resource;
+    }
+
+    /**
+     * Initializes all fields which have a default value set, and which aren't already set.
+     * This method is only meant to be called after new resources are created.
+     */
+    protected void initDefaultValues() {
+        getMembers(Field.class)
+                // discard fields which already have a value set
+                .filter(field -> !field.isSet())
+                // only consider fields which have a default value
+                .filter(field -> field.defaultValue.isPresent())
+                .forEach(field -> field.set(field.defaultValue.get()));
+    }
```

Comment:
Yes I mean big-R (we should coin that term going forward :smile: . 

Right now this method is used when calling `ChildResource.getOrCreate` but in the future it should be invoked if/when we implement a way to create `Resource` objects using this API.

Reply:
Oh, I see. So _today,_ when we instantiate a new Module, it must be based off an existing Resource... _however,_ it is _not_ a requirement that that underlying jcr node must have an existing cachedContent node created - is that right? Is that what this is written for so that the code continues to function correctly?

Reply:
_kind of_ ... you are right that there might not be a `cachedContent` node created and this method will help with that. My intention was bit wider-reaching: I saw this also as a way to remove the current pre-req that a `Resource` needs to exist before being able to wrap it around a `SlingResource` object.

Reply:
Hmm... alright. I'm not sure I can fully follow without really diving and and playing with it, but for now, I am convinced that this will solve our current purposes and if we run into any gotchas or surprises later, we can iterate then.